server/cluster/cluster.go

```diff
@@ -17,33 +18,55 @@ import (
 	clientv3 "go.etcd.io/etcd/client/v3"
 )
 
-type metaData struct {
+type MetaData struct {
 	cluster         *clusterpb.Cluster
 	clusterTopology *clusterpb.ClusterTopology
 }
 
+func (m MetaData) GetCluster() *clusterpb.Cluster {
+	return m.cluster
+}
+
+func (m MetaData) GetClusterTopology() *clusterpb.ClusterTopology {
+	return m.clusterTopology
+}
+
 type Cluster struct {
 	clusterID uint32
 
 	// RWMutex is used to protect following fields.
 	lock         sync.RWMutex
-	metaData     *metaData
-	shardsCache  map[uint32]*Shard  // shard_id -> shard
-	schemasCache map[string]*Schema // schema_name -> schema
-	nodesCache   map[string]*Node   // node_name -> node
+	metaData     *MetaData
+	shardsCache  map[uint32]*Shard         // shard_id -> shard
+	schemasCache map[string]*Schema        // schema_name -> schema
+	nodesCache   map[string]*Node          // node_name -> node
+	shardLockMap map[uint32]*ShardWithLock // shard_id -> shardLock
 
 	storage       storage.Storage
 	kv            clientv3.KV
 	hbstream      *schedule.HeartbeatStreams
-	coordinator   *coordinator
 	schemaIDAlloc id.Allocator
 	tableIDAlloc  id.Allocator
 }
 
+func (c *Cluster) GetNodesCache() map[string]*Node {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.nodesCache
+}
```

Comment:
Such an operation is dangerous because the `nodesCache` will not be protected after returned.

Besides this, it seems some other fields are also used like this, which need fixing too.